rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1272 @@
+and can be ignored in the `git blame` command using [`--ignore-revs-file`](https://www.git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-fileltfilegt).
+
+### Formatting gotchas requiring manual intervention
+
+Some code patterns in Nixpkgs will result in a sub-optimal format,
+because an auto-formatter cannot do exceptions based on context.
+A lot of the times though, the same program can be equivalently expressed in a prettier way.
+
+#### CLI argument flags
+
+It is common to pass in CLI flags (e.g. to builders) like this:
+
+```nix
+[
+  "--some-flag" "some-value"
+]
+```
+
+However, this will be formatted sub-optimally:
+
+```nix
+[
+  "--some-flag"
+  "some-value"
+]
+```
+
+If the CLI also accepts GNU-style flags, a more structured helper can be used instead:
+
+```nix
+lib.cli.toGNUCommandLine { } {
+  some-flag = "some-value";
+}
+```
+
+#### Singleton lists
+
+Sometimes a list only needs a single element
+and there's no expectation to add more in the future.
+This would be formatted like this:
+
+```nix
+{
+  list = [
+    {
+      foo = 10;
+      bar = 20;
+    }
+  ];
+}
+```
+
+In this case one level of indentation can be saved using [`lib.singleton`](https://nixos.org/manual/nixpkgs/stable#function-library-lib.lists.singleton):
+
+```nix
+{
+  list = lib.singleton {
+    foo = 10;
+    bar = 20;
+  };
+}
+```
+
+## Drawbacks
+
+- No automatic code format can be as pretty as a carefully crafted manual one. There will always be "ugly" edge cases.
+  - However, we argue that on average the new format will be an improvement,
+    and that contributors should not waste their time making the code slightly more pretty.
+- Every formatter will have bugs, but the Nix formatter team will be able to make fixes.
+- Having a commit that changes the formatting, can make git blame harder to use.
+  - This can be [worked around in interfaces](#git-blames).
+
+## Alternatives
+
+- Keep the status quo of not having an official formatter.
+  The danger is that this creates discord within the Nix community.
+  The current friction and maintainer churn due to bad formatting may arguably be small, but not negligible.
+- Specify a different format.
+- Specify a format, but don't enforce it in CI, to allow manually tweaking the output if necessary.
+- Apply the format incrementally, i.e. only changed files get formatted, to make a more graceful transition.
+  However, such an approach would not reduce the amount of merge conflicts,
+  and increase the workload for contributors during that time significantly.
+
+## Unresolved questions
+
+## Future work
+
+- General style guidelines beyond AST reformatting
+- Making widespread use of linters like Nixpkgs-hammering
+- Enforcing the format to other official repositories
+
+-----
+
+## Initial standard Nix format
```

Comment:
I didn't know where else to open a thread for this because it is missing: Comments.

Apparently, the WIP formatter turns multi-line comments into single-line comments. This was confirmed to me as intended by @piegamesde.

I'd like for formatting rules regarding multi-line comments aswell as single-line comments to be added.

These should also include rules for where to place comments relative to the expression it refers to. I'd like to forbid placing the comment in the lines after the expression for example. This is also relevant for the auto-formatter because it must keep existing comments in sensible places too.

Reply:
We just added a section to specify that the formatting of strings shouldn't be changed (https://github.com/NixOS/rfcs/pull/166/commits/8cb46fd96dc650b81b28f0d28f08805458d27951), and I'd be in favor of also doing the same for comments.

Though there could be some special cases where multiline comments aren't necessary anymore due to a reformatting, where it might make sense to turn them into single-line comments.

Reply:
I agree that string format should stay as is. More than a handful of times I’ve emptied a string momentarily only for it to be converted.

Reply:
> Apparently, the WIP formatter turns multi-line comments into single-line comments. This was confirmed to me as intended by @piegamesde.

Just to be clear, this only applies to *some* comments. Specifically, to multi-line comments where all intermittent lines start with `*`. I don't know why this rule was added, however a quick test shows that it only affects a couple dozen of files in Nixpkgs anyways: https://github.com/piegamesde/nixpkgs/pull/4/commits/e371bdbbd11ba31976f5e558ffd9d395f3f631e2. (My hypothesis is that adding normalization was easier than getting them to render correctly in all cases)

Nevertheless, both the implementation and the RFC needs some polish here. I propose at least the following rules:

- `/**` comments must be handled according to RFC #145
- Single-line `/*` comments must be converted to `#` comments
- Single-line comments may be moved up or down a line to improve the layout

I think having some more rules which allow some kinds of content normalization, especially to keep things properly aligned, would be interesting to have. But I'd have to gather more examples on that before proposing anything.

Reply:
@piegamesde please also note that there can also be the case where there is a comment *between* two expressions.

It's very common to "declare" dual-licenses like this:

https://github.com/NixOS/nixpkgs/blob/5ef42fcd84b2baa16f43554f1c1f1d614e23ef9a/pkgs/servers/static-web-server/default.nix#L38

What I also sometimes like to do this little trick to add comments to multi-line strings without adding the comment inside the string:

https://github.com/NixOS/nixpkgs/blob/5ef42fcd84b2baa16f43554f1c1f1d614e23ef9a/pkgs/development/libraries/dbus/default.nix#L38

(Though sadly I don't think this will be possible anymore.)

Reply:
Could you please elaborate on the intended mental parsing of the latter example? For someone who's never touched that part of the codebase/are unfamiliar with it?

I couldn't figure out for myself what that comment tried to tell me.

Reply:
@x10an14 it concatenates two multi-line strings. These two are the same value (`"foo\nbar\n"`):

```Nix
''
  foo
  bar
''
```

```Nix
''
  foo
'' + ''
  bar
''
```

But because you're outside the string declaration at the `+`, you can add an in-line comment without changing the value:

```Nix
''
  foo
'' + /* you could write anything here */ ''
  bar
''
```

Once you recognise this pattern, it becomes very easy to read and doesn't break the flow of the string. It's actually quite useful for adding comments to code that otherwise wouldn't allow a comment at such a place as is often the case with bash scripts.

The current style turns it into this:

```Nix
''
  foo
''
# you could write anything here
+ ''
  bar
''
```

There's now 3 lines of code between `foo` and `bar` eventhough there's still only a single newline in the actual string.

You could do

```Nix
''
  foo
'' # you could write anything here
+ ''
  bar
''
```

and that's permitted under the new style but the comment would actually be "connected" to the `foo` string and there's still an additional newline.

Reply:
We could discuss the necessity of that specific comment, and we could also discuss that in this case the comment could equally well be written as part of the string.

But the general pattern shown in this example is real and valid, since it generalized to other cases where such workarounds are not possible.

Have another arbitrary example from Nixpkgs:

```nix
      (with all;
        # disable default openssl extension
        (lib.filter (e: e.pname != "php-openssl") enabled)
        # use OpenSSL 1.1 for RC4 Nextcloud encryption if user
        # has acknowledged the brokenness of the ciphers (RC4).
        # TODO: remove when https://github.com/nextcloud/server/issues/32003 is fixed.
        ++ (if cfg.enableBrokenCiphersForSSE then [ cfg.phpPackage.extensions.openssl-legacy ] else [ cfg.phpPackage.extensions.openssl ])
        ++ optional cfg.enableImagemagick imagick
        # Optionally enabled depending on caching settings
        ++ optional cfg.caching.apcu apcu
        ++ optional cfg.caching.redis redis
        ++ optional cfg.caching.memcached memcached
      )
```

------

@Atemu the issues I have with your proposal are:

- It does not scale well to very long comments or multi-line comments. Comments are never broken up and therefore not part of any line limit calculations, therefore having any tokens (`''`, `[`) after a comment might end up placing them way off-screen.
- It only works for strings, lists and attrsets and does not generalize to other situations. As soon as functions like `lib.optional` are introduced –which is a very common pattern–, having the inline comment would quickly get weird results. The currently proposed handling of comments generalizes well to lots of situations, increasing overall consistency.